woocommerce/woocommerce-gateway-stripe
# Prevent duplicate customer creation during guest checkout

Fixes #3124

If you place an order as a guest, it doesn’t associate the purchases with an existing Stripe customer. Instead, it creates a new customer.

This PR fixes the issue by searching for an existing customer in Stripe before creating a new customer for guest checkouts.

This PR depends on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2980. The current version of Stripe (`'2019-09-09'`) does not support the search API. So for the changes in this PR to work we have to update the Stripe API version first which is addressed in the [mentioned PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2980).

## Changes proposed in this Pull Request:
- Using the customer search API to look up the customer with email and name.
- Before creating a customer in Stripe for a guest user, we are now checking if a customer already exists in the Stripe end for this email and name. If exists, we won't create a new customer.

## Testing instructions
#### Reproduce the issue for guest checkout
1. Checkout to the `develop` branch.
2. As a guest user, add a product to the cart and go to the checkout page.
3. Fill up the details, and use a new email on the checkout form. (i.e, `tom.cat@example.com`)
4. Complete the checkout with the card payment method. 
5. Go to your stripe dashboard and confirm a new user has been created.
6. Now as a guest user, again add a product to the cart and complete the checkout using the same details used in the previous checkout (`tom.cat@example.com`)  
7. Go to your stripe dashboard again and confirm that a second user has been created with the same email.

#### Test for guest checkout
1. Checkout to the `fix/3124-duplicate-customer` branch.
2.  In `includes/class-wc-stripe-api.php` update the `STRIPE_API_VERSION` to `'2023-10-16'`
3. As a guest user, add a product to the cart and go to the checkout page.
4. Fill up the details, and use a new email on the checkout form. (i.e, `jerry.mouse@example.com`)
5. Complete the checkout with the card payment method. 
6. Go to your stripe dashboard and confirm a new user has been created.
7. Now as a guest user, again add a product to the cart and complete the checkout using the same details used in the previous checkout (`jerry.mouse@example.com`)  
8. Go to your stripe dashboard again and confirm that no new user has been created with the same email.
9. In the Stripe dashboard, click the customer and go to the customer details page. Confirm that 2 payments are linked to this customer.

#### Test for logged in user 
No changes has been made related to logged in user, so it should keep working as before.
1. Checkout to the `fix/3124-duplicate-customer` branch.
2. In `includes/class-wc-stripe-api.php` update the `STRIPE_API_VERSION` to `'2023-10-16'`
3. Log in to the store as an existing user and add a product to your cart.
4. Complete the checkout with the card payment method for this logged in user.
5. Go to your stripe dashboard and confirm no new user has been created.
6. In the Stripe dashboard, find the customer and go to the customer details page. Confirm that the above transaction is listed in the payments section of the corresponding customer.
